File: tests/typesafety/typescript/dispatch_typed.ts
// expect: passes

type Status = "draft" | "sent" | "paid";

type Invoice = { readonly number: string; readonly status: Status };

function send(invoice: Invoice): Invoice {
  return { number: invoice.number, status: "sent" };
}

function recordPayment(invoice: Invoice): Invoice {
  return { number: invoice.number, status: "paid" };
}

function keep(invoice: Invoice): Invoice {
  return invoice;
}

const handlers: Record<Status, (invoice: Invoice) => Invoice> = {
  draft: send,
  sent: recordPayment,
  paid: keep,
};

export function advance(invoice: Invoice): Invoice {
  return handlers[invoice.status](invoice);
}

File: tests/typesafety/python/dispatch_typed.py
# expect: passes
# run: yes
# title: The table declares one handler shape, and every entry is held to it
# improves: dispatch_untyped
from collections.abc import Callable
from dataclasses import dataclass
from typing import Literal

type Status = Literal["draft", "sent", "paid"]


@dataclass(frozen=True)
class Invoice:
    number: str
    status: Status


def send(invoice: Invoice) -> Invoice:
    return Invoice(invoice.number, "sent")


def record_payment(invoice: Invoice) -> Invoice:
    return Invoice(invoice.number, "paid")


def keep(invoice: Invoice) -> Invoice:
    return invoice


HANDLERS: dict[Status, Callable[[Invoice], Invoice]] = {
    "draft": send,
    "sent": record_payment,
    "paid": keep,
}


def advance(invoice: Invoice) -> Invoice:
    return HANDLERS[invoice.status](invoice)


assert advance(Invoice("INV-7", "draft")).status == "sent"
assert advance(Invoice("INV-7", "paid")).status == "paid"
